Formula adjusting in context of column shifting.

zmau opened the pull request:
FormulaShifter class is upgraded in a way that now it can also adjust cell references in formulas when columns are shifted. It works simple :
- transpose cell reference;
- adjust it using existing algorithm for adjusting when rows are shifted;
- transpose result back.

XSSFSheet is upgraded. Now there is shiftColumns() methods, and it works similar to shiftRows(). Except formula adjusting, other issues like update of hyperlinks and named ranges are also covered for column shifting. Most of functionality of XSSFRowShifter class is moved to new class XSSFShiftingManager, and is used from both shiftRows() and shiftColumns() methods.

zmau committed:
Formula adjusting in context of column shifting.

zmau committed:
Merge branch 'trunk' into trunk

asfgit commented:
Can one of the admins verify this patch?

pjfanning commented on src/java/org/apache/poi/hssf/usermodel/HSSFRow.java:
```diff
@@ -113,7 +113,27 @@ public HSSFCell createCell(int column)
+     * @return HSSFCell a high level representation of the created cell.
+     * @throws IllegalArgumentException if columnIndex < 0 or greater than 255,
+     *   the maximum number of columns supported by the Excel binary format (.xls)
+     * @deprecated POI 3.15 beta 3
```
the deprecation statement has the wrong version - should be 4.0.0

zmau replied:
removing whole method

pjfanning commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -548,4 +556,105 @@ private static Ptg createDeletedRef(Ptg ptg) {
+	}
+
+	public static SpreadsheetVersion getSpreadsheetVersion(Sheet sheet){
+		if(sheet.getWorkbook() instanceof XSSFWorkbook)
```
could you fix the indentation here (and elsewhere in your code)? 4 spaces per indent

pjfanning commented on src/java/org/apache/poi/ss/usermodel/Row.java:
```diff
@@ -50,6 +50,28 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+     * @see CellType#STRING
+     * @deprecated POI 3.15 beta 3. Use {@link #createCell(int, CellType)} instead.
+     */
+    Cell createCell(int column, int type);
```
This was removed because we no longer want to pass cell type as an int - why do you want to add it back?

zmau replied:
In fact I do not. I used a bit older version as a base for my version, and this was still in charge. 
Removing now, from all *Row classes.

onealj replied:
Please rebase your pull request.

pjfanning commented on src/java/org/apache/poi/ss/usermodel/Sheet.java:
```diff
@@ -702,6 +702,7 @@ Licensed to the Apache Software Foundation (ASF) under one or more
      * @param resetOriginalRowHeight whether to set the original row's height to the default
      */
     void shiftRows(int startRow, int endRow, int n, boolean copyRowHeight, boolean resetOriginalRowHeight);
+    void shiftColumns(int startColumn, int endColumn, int n);
```
this needs javadoc

zmau replied:
ok, adding javadoc

pjfanning commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+        return shiftedRegions;
+    }
+    
+    public static void cloneCellValue(org.apache.poi.ss.usermodel.Cell oldCell, org.apache.poi.ss.usermodel.Cell newCell) {
```
can you `import org.apache.poi.ss.usermodel.Cell` ?

zmau replied:
sure

onealj replied:
See my earlier comment about cloning cells. We can probably shift cells without cloning cells. For example, this implementation doesn't copy over the cell style, conditional formatting, array formula, table cell, or cached formula result, to name a few. Even if we tried to clone a cell, the method would be private.

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/SXSSFRow.java:
```diff
@@ -132,6 +132,24 @@ public SXSSFCell createCell(int column)
+     * @deprecated POI 3.15 beta 3. Use {@link #createCell(int, CellType)} instead.
+     */
+    @Override
+    public SXSSFCell createCell(int column, int type)
```
As per previous comment, I don't understand why this is being added back.

zmau replied:
also removed again

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/SXSSFSheet.java:
```diff
@@ -2115,4 +2115,10 @@ public void setTabColor(int colorIndex){
+    @NotImplemented 
+    @Override 
+    public void shiftColumns(int startColumn, int endColumn, int n){ 
+      throw new RuntimeException("NotImplemented"); 
```
use UnsupportedOperationException

zmau replied:
ok

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFCell.java:
```diff
@@ -53,6 +53,7 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTCellFormula;
 import org.openxmlformats.schemas.spreadsheetml.x2006.main.STCellFormulaType;
 import org.openxmlformats.schemas.spreadsheetml.x2006.main.STCellType;
+import org.apache.poi.xssf.model.CalculationChain;
```
could you review the order in which you add imports?

zmau replied:
ok

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFCell.java:
```diff
@@ -1311,4 +1312,23 @@ void notifyArrayFormulaChanging(){
     }
+    
+    /***
+     * Moved from XSSFRow.shift(). Not sure what is purpose. 
```
this doesn't read like public javadoc - could you use // comments for this sort of comment?

zmau replied:
Sure, did it.
But I'd like someone to review method (it was a snippet inside other method) and write a real comment for it.
Anyway, I'll remove my comment once we're done with this patch.

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFRowShifter.java:
```diff
@@ -60,228 +48,130 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+        formulaShiftingManager.updateConditionalFormatting();
+    }    
+    /**
+        @deprecated, use FormulaShiftingManager.updateHyperlinks() directly instead
```
could you include the 4.0.0 version number in the deprecation messages?

zmau replied:
Ok. I have also added @Deprecated annotations. Now it looks like 

/**
        @deprecated in POI 4.0.0, use FormulaShiftingManager.updateHyperlinks() directly instead
     */ 
    @Deprecated
    public void updateHyperlinks(FormulaShifter shifter) {

Is that fine ?

onealj replied:
The javadoc `@deprecated` is lower case.
The Java Annotation `@Deprecated` is title case.

pjfanning commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class XSSFColumnShifterTest {
```
Can you fix the formatting?

zmau replied:
I do not understand what do you exactly mean.

pjfanning replied:
the code formatting - no tabs, use 4 spaces for indents

zmau replied:
Fixed that, and committed changes.
(Other changes are still not committed, I think it will be on Monday.)

pjfanning commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+		row = sheet1.createRow(5);
+		row.createCell(1, CellType.NUMERIC).setCellValue(1.5);
+		/*
+		 * sheet2 = (XSSFSheet)wb07.createSheet("sheet2"); row =
```
What is the plan for all the commented out test code?

zmau replied:
Now I have removed comment marks.
But I want to know is there a common method for displaying sheets in console in POI. If there is one, I'd like to use that instead of my writeSheetToLog() method. If there is not, would you like me to create it (it should be a bit more subtle than writeSheetToLog()) ?

pjfanning replied:
I don't know of a util like this - we can add one another time if is useful but for now, just keep the write method local to your test.

wirechen commented:
help! :(
Is there any method could get the bottom row number? like the `getTopRow`
https://github.com/apache/poi/blob/trunk/src/java/org/apache/poi/hssf/usermodel/HSSFSheet.java#L1431-L1434

zmau committed:
fixed some details

as discussed on https://github.com/apache/poi/pull/81
# Conflicts:
#	src/java/org/apache/poi/ss/formula/FormulaShifter.java

zmau committed:
fixed some details

as discussed on https://github.com/apache/poi/pull/81

zmau committed:
Merge remote-tracking branch 'origin/trunk' into trunk

zmau committed:
Replaced tab characters with spaces

(Other commented issues are not covered in this commit.)

onealj commented on src/java/org/apache/poi/hssf/usermodel/HSSFCell.java:
```diff
@@ -950,7 +950,7 @@ public HSSFCellStyle getCellStyle()
      */
 
-    protected CellValueRecordInterface getCellValueRecord()
+    public CellValueRecordInterface getCellValueRecord()
```
Is there any reason that `getCellValueRecord()` needs to be public? I didn't see any usages of this method in the diff.

zmau replied:
You are right. There was a need in old version (with CompleteFormulaShifter), now it can be protected again.

pjfanning replied:
can we undo this change and make method protected again?

zmau replied:
Sure. I just did it in local. I have finally fetched Javen's commits, and now I am in the process of checking those problems.

zmau replied:
In my project I have a method which copies ranges of cells. Inside that method I need to call method getCellValueRecord() :

	case EXCEL97 : 
		FormulaRecordAggregate fra = (FormulaRecordAggregate)((HSSFCell)newCell).getCellValueRecord();
		sheetIndex = destSheet.getWorkbook().getSheetIndex(destSheet); 
		ptgs = fra.getFormulaTokens();
		if(adjustBothDirections(ptgs, sheetIndex, deltaX, deltaY))
			fra.setParsedExpression(ptgs);
		break;

 so I would like it to be public.

Other option is to have range copying implemented inside poi. I can try to work that out, although I am not sure when I will find time for that. 
Which option do you prefer ?

onealj commented on src/java/org/apache/poi/hssf/usermodel/HSSFSheet.java:
```diff
@@ -2640,4 +2621,32 @@ public void setActiveCell(CellAddress address) {
+        _workbook.getWorkbook().updateNamesAfterCellShift(shifter); 
+    } 
+
+        public void shiftColumns(int startColumn, int endColumn, int n){ 
```
fix inconsistent whitespace

pjfanning replied:
can you fix the formatting in this file?

zmau replied:
Sure, just done it in local.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+import org.apache.poi.xssf.usermodel.helpers.XSSFRowShifter; 
+ 
+public class HSSFColumnShifter extends ColumnShifter{ 
+    private static final POILogger logger = POILogFactory.getLogger(XSSFRowShifter.class); 
```
`getLogger(HSSFRowShifter.class)`

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+import org.apache.poi.util.POILogger; 
+import org.apache.poi.xssf.usermodel.helpers.XSSFRowShifter; 
+ 
+public class HSSFColumnShifter extends ColumnShifter{ 
```
There isn't anything in this class that is HSSF-specific. All of this code could be in the ColumnShifter class and be reused by XSSFColumnShifter.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+     * (see constructor for parameters) 
+     */ 
+ 
+    private void shiftColumnsRight(){ 
```
`shiftColumnsRight()` and `shiftColumnsLeft()` are very similar (lots of duplicated code). Are 2 methods needed here?

zmau replied:
I prefer it that way because direction of loops and shifting are different. Surely I could introduce some parameters, and get over it, but I think that that single method would be much harder to read and understand.
Anyway, if you really want me to merge them, I'll do it.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+                } 
+                else { 
+                    newCell = row.createCell(columnIndex + shiftStep, oldCell.getCellType()); 
+                    cloneCellValue(oldCell,newCell); 
```
Is there a more performant way to move cells from source column to destination column without copying the cell contents and recreating the cell somewhere else? This seems like it would be slow and cause some problems if there were customizations to that cell that POI didn't know how to clone.

zmau replied:
Yes, it is a problem, but I am not aware of a way to just change cell's position inside a row. Help from people who are more familiar with framework would be fine here.

onealj replied:
Take a look at XSSFRow#createCell and XSSFCell's constructor.
https://svn.apache.org/viewvc/poi/trunk/src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFRow.java?view=markup
There's a CTCell xmlbean that is added to a TreeMap<columnIndex as Integer, cell as XSSFCell> _cells in XSSFRow. Similarly, there's a TreeMap<rowIndex as Integer, row as XSSFRow> _rows in XSSFSheet.
You can change the row or column of that CTCell by moving it into a different location in the _cells or _rows TreeMaps. No need to create a new xmlbean.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+                    cloneCellValue(oldCell,newCell); 
+                    if(columnIndex <= firstShiftColumnIndex + shiftStep - 1){ // clear existing cells on place of insertion 
+                        oldCell.setCellValue(""); 
+                        oldCell.setCellType(CellType.STRING); 
```
If we're trying to clear the original cell, we'll get a smaller workbook if we remove the cell or set the cell ato blank than setting the cell value to an empty string.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+     */ 
+ 
+    private void shiftColumnsRight(){ 
+        for(int rowNo = 0; rowNo <= shiftingSheet.getLastRowNum(); rowNo++) 
```
`for (Row row : sheet)` is more efficient for processing all rows in a sheet.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+            Row row = shiftingSheet.getRow(rowNo); 
+            if(row == null) 
+                continue; 
+            for (int columnIndex = lastShiftColumnIndex; columnIndex >= firstShiftColumnIndex; columnIndex--){ // process cells backwards, because of shifting  
```
While row shifting can be accomplished with a single for loop, column shifting probably requires two nested for loops (unless you can operate over the data structures directly). I'm worried that this column shifting strategy will be very slow for large sheets. If this turns out to be the case, we can add a method to get a `SortedMap` over the rows from `firstShiftColumnIndex` to `lastShiftColumnIndex`. For now, let's assume your code is fast enough.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+                Cell newCell = null; 
+                if(oldCell == null){ 
+                    newCell = row.getCell(columnIndex + shiftStep); 
+                    newCell = null; 
```
Why is `newCell` reassigned to `null`? What is this code trying to do?

zmau replied:
To nullify that cell in sheet. How can I achieve that ?

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+                if(oldCell == null){ 
+                    newCell = row.getCell(columnIndex + shiftStep); 
+                    newCell = null; 
+                    continue; 
```
why `continue`? This code needs some comments to describe what cases it's trying to handle.

onealj commented on src/java/org/apache/poi/hssf/usermodel/helpers/HSSFColumnShifter.java:
```diff
@@ -0,0 +1,102 @@
+            for (int columnIndex = 0; columnIndex < row.getLastCellNum(); columnIndex++){  
+                HSSFCell oldCell = row.getCell(columnIndex); 
+                if(columnIndex >= firstShiftColumnIndex + shiftStep && columnIndex < row.getLastCellNum() - shiftStep){ // shift existing cell  
+                    org.apache.poi.ss.usermodel.Cell newCell = null; 
```
`Cell newCell = null;` will suffice.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -33,14 +35,16 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 public final class FormulaShifter {
 
-    private static enum ShiftMode {
+    public static enum ShiftMode {
```
What's the justification for making `ShiftMode` a public enum? `ShiftMode` is intentionally private since it's an implementation detail of POI. Since `FormulaShifter` is a final class, it's unlikely that `ShiftMode` needs to be accessible or overridden by code outside of POI. If more accessibility is needed, `package-private` or `protected` would be more appropriate choices for access modifiers.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -33,14 +35,16 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 import org.apache.poi.ss.formula.ptg.RefPtg;
 import org.apache.poi.ss.formula.ptg.RefPtgBase;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.xssf.usermodel.XSSFWorkbook;
```
OOXML classes are not allowed in o.a.p.ss.*. Remove this import.

pjfanning replied:
can you make the change that Javen suggested?

zmau replied:
Resolved locally.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -17,6 +17,8 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 package org.apache.poi.ss.formula;
 
+import org.apache.poi.hssf.usermodel.HSSFWorkbook;
+import org.apache.poi.hssf.util.CellReference;
```
The o.a.p.ss.* package should be agnostic to specific implementations (HSSF, XSSF). Remove these imports.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -67,6 +71,9 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 
     private final ShiftMode _mode;
 
+    private boolean _rowModeElseColumn; 
```
How about adding `ColumnMove` to `ShiftMode` instead? `ShiftMode.ColumnMove` is more in the spirit of this class.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -104,16 +112,33 @@ private FormulaShifter(int srcSheetIndex, int dstSheetIndex) {
             SpreadsheetVersion version) {
-        return new FormulaShifter(externSheetIndex, sheetName, firstMovedRowIndex, lastMovedRowIndex, numberOfRowsToMove, ShiftMode.RowMove, version);
+        FormulaShifter instance = new FormulaShifter(externSheetIndex, sheetName, firstMovedRowIndex, lastMovedRowIndex, numberOfRowsToMove, ShiftMode.RowMove, version);
+        return instance;
```
`return new FormulaShifter(`... is fine. No need to store the returned value in `instance`.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -104,16 +112,33 @@ private FormulaShifter(int srcSheetIndex, int dstSheetIndex) {
+                firstShiftItemIndex, lastShiftItemIndex, shiftStep, ShiftMode.RowMove, getSpreadsheetVersion(shiftingSheet));
+        instance._rowModeElseColumn = _rowModeElseColumn;
+        return instance; 
+    }
```
`ItemShift` is too vague to have as part of the public API.
`_rowModeElseColumn` is also a poor name to have as part of a public API. Public API's shouldn't start with underscore, and boolean variables should be named something that inherently has a yes/no response, such as "isRowMode". Since the shifting mode isn't binary (row, column, sheet) x (copy, move), boolean variables are a poor fit here. Enums are a better fit here, which is why we have them.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -104,16 +112,33 @@ private FormulaShifter(int srcSheetIndex, int dstSheetIndex) {
+        instance._rowModeElseColumn = rowModeElseColumn;
+        return instance; 
+    }
+    // maybe should be deprecated, and previous one should be used
```
Prefer `RowCopy` and `ColumnCopy` methods in a static class over one `ItemCopy` method, since `Item` is not binary.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -548,4 +556,105 @@ private static Ptg createDeletedRef(Ptg ptg) {
+     * @param currentExternSheetIx - the extern sheet index of the sheet that contains the formula being adjusted
+     * @return <code>true</code> if a change was made to the formula tokens
+     */
+    public boolean adjustFormula(Ptg[] ptgs, int currentExternSheetIx) {
```
Wrong level of abstraction. `adjustFormula` doesn't directly care if we're modifying the row or column portion of a cell reference. That's what `adjustPtg` does, using the `ShiftMode` to determine how to modify the cell reference.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -548,4 +556,105 @@ private static Ptg createDeletedRef(Ptg ptg) {
+        return sheet.getWorkbook().getSheetIndex(sheet);
+    }
+
+    public static SpreadsheetVersion getSpreadsheetVersion(Sheet sheet){
```
This method has been defined in other classes. Don't repeat that logic here.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -548,4 +556,105 @@ private static Ptg createDeletedRef(Ptg ptg) {
+    }
+
+    public static SpreadsheetVersion getSpreadsheetVersion(Sheet sheet){
+        if(sheet.getWorkbook() instanceof XSSFWorkbook)
```
o.a.p.ss.* shouldn't reference specific implementations. This code doesn't work for SXSSFWorkbook, future implementations, or derivative works. Because this class is final, users can't override this static method to have the correct behavior if they implement their own *SS*Workbook class.

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+
+public class ColumnShifter {
+    protected final Sheet shiftingSheet;
+    protected FormulaShifter shifter;
```
Can these both be `private final`?

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+import org.apache.poi.ss.util.CellRangeAddress;
+
+public class ColumnShifter {
+    protected final Sheet shiftingSheet;
```
The variable name `sheet` is good enough. It's obvious that any shift inside the class `ColumnShifter` has something to do with shifting. Making it `private` would make that relationship even clearer.

zmau replied:
At first I named that variable just sheet. But later, when testing adjusting of cross-sheet references in formulas, I caught myself being confused, and having problem making difference between sheet which is shifted, and sheet which contains cell with formula. 
So please, try to think about it once again, and think from less-poi-experienced-developer's point of view.

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+
+    public ColumnShifter(Sheet sheet, FormulaShifter shifter) {
+        shiftingSheet = sheet;
+        this.shifter = shifter;
```
Consistency: either use `this.` for both or neither. With mixed usage, my brain tells me that the non-`this` variable is somehow different--either a static (class) or local variable.

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+package org.apache.poi.ss.usermodel.helpers;
```
This file will need the Apache ASL 2.0 header in order to be merged.

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    public List<CellRangeAddress> shiftMergedRegions(int startColumnIndex, int endColumnIndex, int n) {
+        List<CellRangeAddress> shiftedRegions = new ArrayList<CellRangeAddress>();
+        Set<Integer> removedIndices = new HashSet<Integer>();
```
Since you started working on this pull request, we have dropped support for Java 6 and 7. That means that the diamond operator is now available (and preferred).
`Set<Integer> removedIndices = new HashSet<>();`

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+        for (int i = 0; i < size; i++) {
+            CellRangeAddress merged = shiftingSheet.getMergedRegion(i);
+
+            // remove merged region that overlaps shifting
```
There might be some ways to improve performance and/or readability here. `CellRangeAddress.containsColumn()` might work here.

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -0,0 +1,96 @@
+            }
+        }
+        
+        if(!removedIndices.isEmpty()) {
```
What happens if you pass an empty set to `removeMergedRegions()`?

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/RowShifter.java:
```diff
@@ -36,11 +36,17 @@ Licensed to the Apache Software Foundation (ASF) under one or more
-    }
-
-    /**
+    protected FormulaShifter shifter; 
```
`final`? `private`?

This variable is never used within this class. Is it only accessed statically? Can the HSSFRowShifter and XSSFRowShifters define their own shifter variable instead? Do those classes even need a shifter variable?

This hasn't been needed before. Why is it needed now?

The variable name `shifter` is not descriptive enough for a `FormulaShifter` instance inside of a `RowShifter` class.

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/RowShifter.java:
```diff
@@ -36,11 +36,17 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+ 
+    public RowShifter(Sheet sh, FormulaShifter shifter) { 
+        sheet = sh; 
+        this.shifter = shifter; 
```
`this` and non-`this`.

Why does the RowShifter need to be passed a FormulaShifter instance? If we develop a MergedRegionShifter and a ConditionalFormattingShifter, will we need to add those variables to the constructor?

Why are both `RowShifter(Sheet sh)` and `RowShifter(Sheet sh, FormulaShifter shifter)` available? How does the code work if the first constructor is called?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFRow.java:
```diff
@@ -113,7 +114,7 @@ public XSSFSheet getSheet() {
     @Override
     @SuppressWarnings("unchecked")
-	public Iterator<Cell> cellIterator() {
+    public Iterator<Cell> cellIterator() {
```
Thanks. I'll fix these whitespace inconsistencies. We're pretty inconsistent with whitespace. 
Eventually we'll run some tools against our codebase to make our whitespace consistent.
If it doesn't relate to your changes, you can leave them out of your pull request.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFRow.java:
```diff
@@ -206,6 +207,26 @@ public XSSFCell createCell(int columnIndex) {
+     * @param type - the cell's data type
+     * @return XSSFCell a high level representation of the created cell.
+     * @throws IllegalArgumentException if the specified cell type is invalid, columnIndex < 0
+     *   or greater than 16384, the maximum number of columns supported by the SpreadsheetML format (.xlsx)
```
Same comment as before (integer cell type was removed). Make sure you've diffed your changes against the base that you edited from, then rebase to trunk.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFRow.java:
```diff
@@ -552,25 +573,13 @@ public String toString(){
      * @param n the number of rows to move
      */
-    protected void shift(int n) {
+    public void shift(int n) {
```
Does `protected` work?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFSheet.java:
```diff
@@ -2985,7 +2987,70 @@ public void shiftRows(int startRow, int endRow, int n) {
+        shiftingManager.updateHyperlinks();
+
+        //rebuild the _rows map
+        Map<Integer, XSSFRow> map = new HashMap<Integer, XSSFRow>();
```
`HashMap<>`

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFSheet.java:
```diff
@@ -2985,7 +2987,70 @@ public void shiftRows(int startRow, int endRow, int n) {
+    }
+    
+    // remove all rows which will be overwritten
+     private void removeOverwritten(XSSFVMLDrawing vml, int startRow, int endRow, final int n){
```
why is a `XSSFVMLDrawing` parameter needed here?

zmau replied:
It is used in the method body. 
I do not understand the logic of that using, I have just moved existing snippet into new method, and did not study the details. I just guess that old existing code has its sense.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFVMLDrawing.java:
```diff
@@ -264,7 +264,7 @@ protected CTShape newCommentShape(){
      * @return the comment shape or <code>null</code>
      */
-    protected CTShape findCommentShape(int row, int col){
+    public CTShape findCommentShape(int row, int col){
```
Does `protected` not work? `CTShape` is an xmlbean class that's an implementation detail of POI that shouldn't be relied upon.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+package org.apache.poi.xssf.usermodel.helpers;
```
ASL 2.0 license header

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import org.apache.poi.ss.ITestDataProvider;
```
`ITestDataProvider` is part of the unit test suite and not included in any of the poi jars. This class will be unrunnable outside of a development environment.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+
+public class XSSFColumnShifter extends ColumnShifter{
+    
+    private static final POILogger logger = POILogFactory.getLogger(XSSFRowShifter.class);
```
`getLogger(XSSFColumnShifter.class)`.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+    
+    private int firstShiftColumnIndex; 
+    private int lastShiftColumnIndex; 
+    private int shiftStep;
```
`shiftStep` is redundant name.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+    private int lastShiftColumnIndex; 
+    private int shiftStep;
+    
+    private XSSFShiftingManager formulaShiftingManager;
```
Can any of these member variables be `final`?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+            shiftColumnsRight();
+        else if(shiftStep < 0)
+            shiftColumnsLeft();
+//          formulaShiftingManager.updateFormulas();
```
Is updating formulas still a work in progress?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+        this.firstShiftColumnIndex = firstShiftColumnIndex;
+        this.lastShiftColumnIndex = lastShiftColumnIndex;
+        this.shiftStep = step;
+        if(shiftStep > 0)
```
Are the variables `firstShiftColumnIndex`, `lastShiftColumnIndex`, and `shiftStep` used after `shiftColumns` is called? If not, these variables should be passed to `shiftColumnsRight` or `shiftColumnsLeft`, eliminating the need to store these variables. Eliminating unnecessary member variables makes the class easier to read (the class less closesly resembles a state machine) and slightly reduces the class's memory consumption.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+     */
+
+    private void shiftColumnsRight(){
+        for(int rowNo = 0; rowNo <= shiftingSheet.getLastRowNum(); rowNo++)
```
Same comment as before. This code looks pretty similar to the HSSF version of shiftColumnsRight and shiftColumnsLeft. For the HSSF and XSSF versions of shiftColumns{Right|Left}, if these can be consolidated into a single function, we're looking at a 75% reduction in code.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+    public void shiftComments(XSSFVMLDrawing vml, int startColumnIndex, int endColumnIndex, final int n, CommentsTable sheetComments){
+        SortedMap<XSSFComment, Integer> commentsToShift = new TreeMap<XSSFComment, Integer>(new Comparator<XSSFComment>() {
+            @Override
+            public int compare(XSSFComment o1, XSSFComment o2) {
```
Why do you need to override `XSSFComment`'s compare function here? Why not implement it in `XSSFComment` directly?

onealj replied:
Disregard. This code is pre-existing, moved from XSSFSheet.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+    }
+    
+    public void shiftComments(XSSFVMLDrawing vml, int startColumnIndex, int endColumnIndex, final int n, CommentsTable sheetComments){
+        SortedMap<XSSFComment, Integer> commentsToShift = new TreeMap<XSSFComment, Integer>(new Comparator<XSSFComment>() {
```
Why do you need to keep your XSSFComments in a SortedMap to be able to shift them? Does shifting order matter?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+    
+    public void shiftComments(XSSFVMLDrawing vml, int startColumnIndex, int endColumnIndex, final int n, CommentsTable sheetComments){
```
`XSSFVMLDrawing` isn't in the signature of most public POI methods. Is it needed here, or can you determine what `vml` is inside this method?

Can `shiftComments` be `private`, `package-private`, or `protected`?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java:
```diff
@@ -0,0 +1,170 @@
+                    commentsToShift.put(xssfComment, newColumnIndex);
+                }
+            }
+            for(Map.Entry<XSSFComment, Integer> entry : commentsToShift.entrySet()) 
```
Do you need to accumulate comments into a commentsToShift map and then process that map? Can you modify the comments as you iterate over them, saving you from having to iterate over the comments twice?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFRowShifter.java:
```diff
@@ -60,228 +48,130 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+    // do the actual moving and also adjust comments/rowHeight
+    // we need to sort it in a way so the shifting does not mess up the structures, 
+    // i.e. when shifting down, start from down and go up, when shifting up, vice-versa
+    public void doShiftingAndProcessComments(XSSFVMLDrawing vml, int startRow, int endRow, final int n, 
```
`private`

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+package org.apache.poi.xssf.usermodel.helpers;
```
ASL 2.0

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+
+public class XSSFShiftingManager {
+
+    private static final POILogger logger = POILogFactory.getLogger(XSSFRowShifter.class);
```
`XSSFShiftingManager.class`

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+import org.openxmlformats.schemas.spreadsheetml.x2006.main.CTWorksheet;
+import org.openxmlformats.schemas.spreadsheetml.x2006.main.STCellFormulaType;
+
+public class XSSFShiftingManager {
```
extends or implements some base/abstract `ShiftingManager` or ShiftingManagerInterface`?

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+    @Internal
+    public void updateRowFormulas(Row row) {
+        for (Cell c : row) {
+               updateCellFormula(row, (XSSFCell) c);
```
ClassCastException will occur if Row is not an XSSFRow.

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+
+    private void updateSheetFormulas(Sheet sh) {
+        for (Row r : sh) {
+            XSSFRow row = (XSSFRow) r;
```
ClassCastException

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+        }
+    }
+    
+    public void updateCellFormula(Row row, XSSFCell cell){
```
ClassCastException

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+        Workbook wb = sheet.getWorkbook();
+        int sheetIndex = wb.getSheetIndex(sheet);
+        final int rowIndex = row.getRowNum();
+        XSSFEvaluationWorkbook fpb = XSSFEvaluationWorkbook.create((XSSFWorkbook) wb);
```
ClassCastException

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+
+    
+    public void updateConditionalFormatting() {
+        XSSFSheet xsheet = (XSSFSheet) shiftingSheet;
```
ClassCastException

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFShiftingManager.java:
```diff
@@ -0,0 +1,286 @@
+    private static final POILogger logger = POILogFactory.getLogger(XSSFRowShifter.class);
+
+    protected final Sheet shiftingSheet;
+    protected FormulaShifter shifter;
```
`private final`?

onealj commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+package org.apache.poi.xssf.usermodel.helpers;
```
Rename to TestXSSFColumnShifter.java
If you've written your unit test generically enough, you can have a single unit test, `BaseTestColumnShifter`, that can be used to test both the `HSSFColumnShifter` and `XSSFColumnShifter`. As a bonus, anything tested in the base unit test will guarantee that `HSSFColumnShifter` and `XSSFColumnShifter` behave the same.

ASL 2.0 header

onealj commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.slf4j.Logger;
```
Is org.slf4j.Logger needed here? We use POILogger for our logging, which has some features that slf4j doesn't.

onealj commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+
+public class XSSFColumnShifterTest {
+
+    //private static Logger log = LoggerFactory.getLogger(XSSFColumnShifterTest.class + "_T");
```
remove commented out code if not needed

onealj commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+        read.close();
+    }
+
+    private void createHyperlink(CreationHelper helper, Cell cell, HyperlinkType linkType, String ref) {
```
Can this be part of your initial workbook?

onealj commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+
+    @Before
+    public void init() {
+        wb07 = new XSSFWorkbook();
```
If you want, you can load a test spreadsheet from test-cases/spreadsheet, saving you the setup code here.

onealj replied:
`wb` is sufficient.

onealj commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+                                                        // for 2 columns, so now
+                                                        // c5 should be null
+        assertEquals(c6Null, null);
+    }
```
A few more test cases:

- shift left
- shift cells left of the first column (should throw some kind of exception, such as IllegalStateException or IllegalArgumentException)
- shift cells right of the last column
- Formulas appear to be well tested. How about numbers, strings, dates, cell styles?

onealj commented on src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java:
```diff
@@ -0,0 +1,384 @@
+        wb.close();
+    }
+
+}
```
Every new class needs a unit test. Every new method needs a unit test.

Ideally the unit tests should be written before implementing anything, which helps both with good API design (focuses your thinking on "if I were a user, how would I like to use this class" rather than "knowing the current implementation, what are the fewest modifications I can make to get something that sorta works, even if it's not the most intuitive or easy to use") as well as makes sure that you've implemented everything correctly--including corner cases.

Were you able to run the entire unit test suite plus XSSFColumnShifterTest against your changes? Did everything pass?

onealj reviewed (changes requested):
Needs more unit tests.

Exposes too much in the public API (use private whenever possible).

Avoid member variables unless they're needed to maintain state between subsequent calls. There were some member variables that looked like they were being used only to allow private methods to be called without any arguments.

Make sure your changes are against the correct version, and rebase your changes to trunk to avoid the `int cell type` changes.

There's more work that is needed to harmonize HSSF and XSSF to avoid code duplication. Leverage interfaces, abstract classes, and base classes as much as possible to reduce duplicated code.

You have identified several methods that are pretty long in POI that could be broken down into independent private methods. I like this proposition and I'll work on merging these changes into POI.

onealj commented on src/java/org/apache/poi/ss/formula/FormulaShifter.java:
```diff
@@ -548,4 +556,105 @@ private static Ptg createDeletedRef(Ptg ptg) {
+                }
+            }
+            else {
+                Ptg transposedPtg = transpose(ptgs[i]);
```
While `column shift = transpose ∘ row shift ∘ transpose` works, it's an inefficient way to do it.

zmau replied:
It is, for large enough sheets. But faster solution would must dive deep into adjusting logic, so it would be much harder to implement, and much more error prone. So - at least for this time - it is out of scope of my work.

onealj replied:
I added FormulaShifter$ShiftMode#ColumnMove to trunk in a recent commit. It was a pretty trivial addition. Unfortunately, the changes required more code duplication than what other languages can offer, but that's Java.

onealj commented:
Most of the ClassCastExceptions are pre-existing issues in the code. Disregard my CCE comments. The CCE's can be fixed later outside of this PR.

onealj commented:
I implemented most of the changes to FormulaShifter and the Row/ColumnShifter helpers. The remaining work (which is the hard part) is updating XSSFRow and HSSFRow, and writing unit tests.

zmau committed:
Merge remote-tracking branch 'origin/trunk' into trunk

pjfanning commented:
Can we get the merge conflicts resolved?

zmau committed:
Merge branch 'trunk' into trunk

zmau committed:
Merge remote-tracking branch 'origin/trunk' into trunk

pjfanning commented on .gitignore:
```diff
@@ -46,3 +46,4 @@ sonar/*/target
 build
 dist
 lib/
+.project
```
could you remove this change?

pjfanning commented on .project:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <projectDescription>
-	<name>ApachePOI</name>
+	<name>ApachePOIFork</name>
```
could you undo this change?

pjfanning commented on src/java/org/apache/poi/hssf/usermodel/HSSFRow.java:
```diff
@@ -113,7 +113,7 @@ public HSSFCell createCell(int column)
         return this.createCell(column,CellType.BLANK);
     }
-
+    
```
Could you undo this change?

zmau replied:
just undone in local

zmau committed:
Merged version from original poi. Fixed comments shifting. And more.

Merged version from original poi trunk, and resolved arising problems.
Fixed issues with comments-shifting while column shifting.
Changed case with shifting columns to negative index. Now it raises exception.
Introduced ColumnShifter and RowShifter instead of ShiftingManager.
Added column-shifting methods again.
Changed project's name back to ApachePOI

zmau committed:
Merged version from original poi trunk, and resolved arising problems.

Fixed issues with comments-shifting while column shifting.
Changed case with shifting columns to negative index. Now it raises exception.
Introduced ColumnShifter and RowShifter instead of ShiftingManager.
Added column-shifting methods again.
Changed project's name back to ApachePOI

zmau commented:
Pushed.
Next to be done :
- refactor cell shifting to avoid cloning;
- remove class ShiftingManger;
- make more test cases inside TestXSSFSheetShiftColumns class.

zmau committed:
Merge branch 'trunk' into trunk

zmau commented:
I wanted to remove my class XSSFShiftingManager, because new class XSSFRowColShifter contains all of its functionality. But I see XSSFShiftingManager is still used inside XSSFRow, and XSSFRowColShifter can not be used there because it is not public. Is there a reason for that ? Do you want to keep class XSSFShiftingManager just for that one call at XSSFRow (line 619) ?

onealj commented:
Correct. XSSFRowColShifter was inspired by your XSSFShiftingManager. The
name is more specific to what it does.
I did not make any changes to XSSFRow. Please replace your usage of
XSSFShiftingManager with XSSFRowColShifter in XSSFRow in your pull request.

zmau commented:
If I have sheets sheet1 and sheet2 on a workbook, and a formula Sheet1!A4+Sheet1!C2 somewhere inside sheet2, should POI match string **Sheet1** from formula, and sheet name **sheet1** ?

I have checked, MS excel keeps its sheet names case insensitive (gui takes care they are always first-letter-upcase-all-others-low), while POI does not match such names (compares with equals() method), and thus does not adjust such a formula while shifting on sheet1.
Is this behavior OK ?

onealj commented:
Correct. Sheet names and formulas are case insensitive in Microsoft Excel.

POI should use String#equalsIgnoreCase(String) to compare sheet names. If
POI uses case sensitive equals to compare sheet names, you found a bug.

onealj commented on src/ooxml/testcases/org/apache/poi/xssf/usermodel/TestXSSFSheetShiftColumns.java:
```diff
@@ -0,0 +1,433 @@
+                for(int columnIndex = 0; columnIndex < 5; columnIndex++){
+                    //String comment = sheet.getCellComment(new CellAddress(rowIndex, columnIndex)) == null ? "no comment" : sheet.getCellComment(new CellAddress(rowIndex, columnIndex)).getString().getString();
+                    //line +=  String.format("; %1$12s    %2$20s",  ""/*getValue(row.getCell(columnIndex))*/, comment);
+                    line +=  String.format("; %1$12s",  getValue(row.getCell(columnIndex)));
```
Use `DataFormatter#formatCellValue(Cell)` instead of writing your own switch statement.
https://poi.apache.org/apidocs/org/apache/poi/ss/usermodel/DataFormatter.html#formatCellValue-org.apache.poi.ss.usermodel.Cell-

There's a working example here: https://poi.apache.org/spreadsheet/quick-guide.html#Getting+the+cell+contents

onealj commented on src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java:
```diff
@@ -119,4 +122,87 @@ private boolean removalNeeded(CellRangeAddress merged, int startColumn, int endC
+        }
+    }
+    
+    private void nullifyCell(Cell cell){
```
I think `Row.removeCell(Cell)` will accomplish what you want here. It will completely remove a cell in the sheet. This includes value, style, borders, etc.

But I don't think we need to remove a Cell, just move the cell to a different column in the Row. It looks like the Row and Cell API don't have methods to move a cell from one column to another. Could you propose an API that would allow you to move a cell to a new column within a row so that you do not have to copy the cell and delete the original?

zmau replied:
Meanwhile I have refactored shifting, and I have resolved this in the other way. You can see it in new commit.

onealj commented on src/java/org/apache/poi/hssf/usermodel/HSSFSheet.java:
```diff
@@ -2659,4 +2660,30 @@ public void setActiveCell(CellAddress address) {
         _sheet.setActiveCellRow(row);
         _sheet.setActiveCellCol(col);
     }
+    private void shiftFormulas(int startRow, int endRow, int n, boolean forRowElseColumnShift){ 
```
Per your suggestion, I moved the this code from HSSFSheet#shiftRows into its own method, `private void updateFormulasForRowShift(int startRow, int endRow, int n)`

A column shifting variant of this private method is needed.

The best place for column shifting code in HSSFSheet is right below the rowShift and helper methods.

zmau replied:
I made common method 
private void updateFormulasForShift (FormulaShifter formulaShifter)
used for both row- and column- shifting, depending on formulaShifter argument.

onealj commented on src/java/org/apache/poi/hssf/usermodel/HSSFSheet.java:
```diff
@@ -2659,4 +2660,30 @@ public void setActiveCell(CellAddress address) {
+        int sheetIndex = _workbook.getSheetIndex(this);
+        String sheetName = _workbook.getSheetName(sheetIndex);
+        short externSheetIndex = _book.checkExternSheet(sheetIndex);
+        FormulaShifter shifter = FormulaShifter.createForRowShift(
```
`createForColumnShift`

onealj commented on src/java/org/apache/poi/hssf/usermodel/HSSFSheet.java:
```diff
@@ -2659,4 +2660,30 @@ public void setActiveCell(CellAddress address) {
+        _workbook.getWorkbook().updateNamesAfterCellShift(shifter); 
+    } 
+
+    public void shiftColumns(int startColumn, int endColumn, int n){ 
```
Add a Javadoc to indicate what features are shifted and what isn't (yet)
Add a `@Beta` annotation to indicate this is still being developed.
Add `@Override`

onealj commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFCell.java:
```diff
@@ -1311,4 +1312,22 @@ void notifyArrayFormulaChanging(){
     }
+    
+    
+    //Moved from XSSFRow.shift(). Not sure what is purpose. 
```
`protected void XSSFRow#shift(int n)` is called by XSSFSheet#shiftRows.
The code does 3 things as part of a row shift

1. Update array formulas (advanced feature; don't worry about this for column shifting yet)
2. Update the calculation chain (advanced feature; likewise)
3. Updates every CTCell's address due to a row shift via `ctCell.setR(r);`
In sheet#.xml, this looks like `<c r="A1" t="n" v="15" />`, where c means cell, r means reference, v means value, and t means type. Failing to update setR would mean that the cell would never get moved.

Are `getRowIndex()` and `getColumnIndex()` updated to the new row and column indices by the time this method is called? It looks like `XSSFRow#shift` was responsible for updating this.
My guess why this code wasn't moved into XSSFCell was because of the number of variables that would have to be passed in as arguments or recomputed by XSSFCell (this is a tight loop: recomputing sheet index and calculation chain for every cell in the shift range is prohibitively expensive).

If you have your development environment set up to run unit tests, you probably would have seen your change fail the row shifting unit tests, which would give some insight into what this code is accomplishing.

XSSFRow.java:
```
protected void shift(int n) {
    int rownum = getRowNum() + n;
    for (Cell c : this) {
        ...
        String r = new CellReference(rownum, cell.getColumnIndex()).formatAsString();
        ...
    }
    setRowNum(rownum);
}
```

zmau committed:
Refactored column shifting

It's not generic cloning solution any more. In XSSF it is now real shifting. In HSSF and SXSSF it is not implemented yet.
Also :
- added checking of cross-sheet references, styles, strings and boolean values;
- removed XSSFShiftingManager.

zmau committed:
Merge remote-tracking branch 'origin/trunk' into trunk

zmau commented:
I executed old unit tests on my fork.
Testing results are all **same as in original version**:
src/testcases 5 errors and 3 failures
src/ooxml/testcases 5 failures (all in class TestUnfixedBugs)
src/scrachpad/testcases	1 failure
src/integrationtest 4 errors

zmau committed:
HSSF column shifting

Improved HSSFSheet.shiftColumns().
Implemented column shifting for HSSFRow .
Made content of TestXSSFSheetShiftColumns generic - now it works for both HSSF and XSSF.
Created ColumnShifting test classes for both XSSF and HSSF variations.

zmau committed:
Added JUnit test for new cell shifting methods XSSFRow and HSSFRow.

Also added some input validations and some javadoc.

zmau committed:
Merge remote-tracking branch 'refs/remotes/origin/trunk'

Conflicts:
	src/java/org/apache/poi/hssf/usermodel/HSSFSheet.java
	src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java
	src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFRow.java
	src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFSheet.java
	src/ooxml/testcases/org/apache/poi/xssf/usermodel/TestXSSFSheetShiftColumns.java
	src/testcases/org/apache/poi/ss/usermodel/BaseTestSheetShiftColumns.java
	src/testcases/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifterTest.java

Commit:
Merge branch 'upstream-trunk' into zmau-trunk

pjfanning commented on src/testcases/org/apache/poi/ss/usermodel/BaseTestSheetShiftColumns.java:
```diff
@@ -0,0 +1,432 @@
+import org.apache.poi.ss.util.CellRangeAddress;
+import org.apache.poi.ss.util.CellUtil;
+import org.apache.poi.util.IOUtils;
+import org.apache.poi.xssf.XSSFITestDataProvider;
```
this doesn't compile - the xssf classes are in poi-ooxml and the poi-ooxml depends on poi (but not vice versa)

zmau replied:
I was not aware of this. It's all one project and it compiles well on my machine.

Anyway, today I have fixed the issue, and separated xssf from ss. Now there are no xssf references in ss class BaseTestSheetShiftColumns.
Now there is new class TestXSSFSheetShiftColumns, and BaseTestSheetShiftColumns became abstract. 
TestXSSFSheetShiftColumns works fine, all tests pass. 
In TestHSSFSheetShiftColumns some tests fail, and that's because I did not implement everything about shifting for hssf scenario (and, as I told you before, I do not plan to implement it for now). So that class can be removed, or it can wait for someone to finish implementation.

pjfanning commented:
@zmau - if this PR is no longer active, could you close it?

zmau committed:
removed dependencies from ss to xssf

Compeletely separated HSSF and XSSF versions of test. XSSF test works good. HSSF test does not, since I did not implement everything for hssf.

Commit:
Merge branch 'trunk' of https://github.com/zmau/poi into zmau-trunk

zmau commented:
I do not want to just close it. I have finished what I had in plan, and I am waiting for someone to review it, and - hopefully - integrate it into poi trunk.
I am glad I saw your reaction, and I am ready to take part in final phase of this project.

pjfanning commented:
Can you raise a new bugzilla issue to fix the failing hssf tests and mark them with junit `@Ignore` annotations in the code?
eg `@Ignore("see <Bugzilla URL>")`

zmau commented:
I can not add annotations to hssf test methods, because all test methods are inside base class BaseTestSheetShiftColumns, since they cover both hssf and xssf.
Anyway I can post on bugzilla if you still want me to.

zmau committed:
made a temporal workaround to make TestHSSFSheetShiftColumns tests pass

Made empty override versions in TestHSSFSheetShiftColumns for all BaseTestSheetShiftColumns methods which fail for hssf.
After appropriate hssf code is implemented, those methods should be removed, in order to re-enable original test methods from BaseTestSheetShiftColumns class.

pjfanning commented:
you can override the implementations with empty implementation (in the subclass)
the tests are run in CI mode, so we can't merge failing tests

zmau commented:
I have made and pushed such a workaround. Bugzilla item is here : https://bz.apache.org/bugzilla/show_bug.cgi?id=62030

zmau committed:
added a reference to bugzilla task

Commit:
Merge branch 'trunk' of https://github.com/zmau/poi into zmau-trunk

pjfanning commented on src/testcases/org/apache/poi/ss/usermodel/BaseTestRow.java:
```diff
@@ -466,4 +466,71 @@ public void testRowStyle() throws IOException {
+        row.createCell(5, CellType.NUMERIC).setCellValue(5);//F
+        row.createCell(6, CellType.NUMERIC).setCellValue(6);//G
+        try{
+            row.shiftCellsLeft(6, 4, 2); // range [6-4] is illegal
```
shouldn't this test be using shiftCellsRight?

pjfanning replied:
this test fails for TestSXSSFCell (likewise for shiftLeft test) - they are inherited from this base class and SXSSF throws NotImplemented exceptions for this functionality

zmau replied:
I guess you're talking about TestSXSSFRow, not TestSXSSFCell.
I see it. Do you want me to do the same thing as for previous issue with HSSF ? 
I've never worked with SXSSF.

pjfanning replied:
If you could override the tests in TestSXSSFRow to stop the shift cell tests from running

zmau replied:
Just pushed that (https://github.com/zmau/poi/commit/40a0f1b595cf2a0bdeb26d56194c0b9bac31dbde).

zmau committed:
Merge pull request #2 from pjfanning/zmau-trunk

Rebase with apache/poi trunk

zmau commented:
Inside my project I have a class RangeCopier with method  which copies content from one sheet-range to another. To be precise, here's javadoc :
	/** Uses input pattern to tile destination region, overwriting existing content. Works in following manner : 
	 * 1.Start from top-left of destination.
	 * 2.Paste source but only inside of destination borders.
	 * 3.If there is space left on right or bottom side of copy, process it as in step 2. 
	 * @param tilePatternRange
	 * @param tileDestRange
	 */
	public void copyRange(Range tilePatternRange, Range tileDestRange)...

Do you think this class should be integrated into POI ? I can do it if it makes sense.

pjfanning commented:
RangeCopier sounds useful. Could you do that in a new Pull Request?

zmau commented:
I have just created a pull request for RangeCopier : https://github.com/apache/poi/pull/94
